### DISCUSSION TITLE
Can't install Ruby gem after publishing to Github Package Registry

### DISCUSSION THREAD

**Discussion by jer-k** [community/community#25162] [Packages]
<p>I’m using the standard template (<a href="https://github.com/actions/starter-workflows/blob/master/ci/gem-push.yml" rel="nofollow noopener">https://github.com/actions/starter-workflows/blob/master/ci/gem-push.yml</a>) for publishing a Ruby gem to the package registry. I commented out the part of publishing to RubyGems, since this gem is private for now. After successfully publishing the gem, I am unable to install it locally. The error I’m getting is a 401 after copying the install command from the packages page.</p>
<pre><code>gem install my_gem --version "0.1.0" --source "https://rubygems.pkg.github.com/jer-k"
ERROR: Could not find a valid gem 'my_gem' (= 0.1.0), here is why:
          Unable to download data from https://rubygems.pkg.github.com/jer-k/ - bad response Unauthorized 401 (https://rubygems.pkg.github.com/jer-k/specs.4.8.gz)
</code></pre>
<p>Some notes:</p>
<p>The repository ‘my_gem’ is a private repo I mantain. </p>
<p>The OWNER environment variable in the action is jer-k</p>
<p>The GPR_AUTH_TOKEN environment variable has package:write package:read repo permissions.</p>
<p>I added the same token to my .gem/credentials following the instructions from <a href="https://help.github.com/en/articles/configuring-rubygems-for-use-with-github-package-registry#authenticating-to-github-package-registry" rel="nofollow noopener">https://help.github.com/en/articles/configuring-rubygems-for-use-with-github-package-registry#authenticating-to-github-package-registry</a></p>
<p>I’m must be missing something about how to authenticate. I was also trying to search how to distinguish between publishing public and private packages (this is a feature the package registry talks about) but I couldn’t find anything to indicate how to do it; maybe that is based on whether or not the repository is private?</p>

**Comment by ethomson** [comment]
<p>Sorry to hear that.  Just to clarify a few things:  you’re using the same GPR_AUTH_TOKEN that you set up as a secret in your Actions workflow?  When you created the file, did you use a text editor?  Are you on Windows?  (That may change the line endings).</p>
<p>If you open bash (Git Bash if you’re on Windows) and run:</p>
<pre><code>printf -- "---\n:github: Bearer ${GPR_AUTH_TOKEN}\n" &gt; $HOME/.gem/credentials
</code></pre>
<p>Is there any change in behavior?</p>

**Comment by jer-k** [comment]
<p>Yes, I’m using the same token in the action and the command line. I’m on OSX and I had set it up through vim originally. I just used that command and the result was the same.</p>

**Comment by palkan** [comment]
<p>Having the same issue (Linux).</p>
<p>~/.gem/credentials is configured properly.</p>
<p>The same token works with bundler on the same machine, but <code>gem install xyz --source \&lt;github\&gt;</code> fails with <code>Unauthorized 401 ([https://rubygems.pkg.github.com/ruby-next/specs.4.8.gz](https://rubygems.pkg.github.com/ruby-next/specs.4.8.gz))</code>.</p>

**Comment by michaelwebb76** [comment]
<p>Also having the same issue, but only for some gems. Bundle install / update simply refuses to find a gem which is clearly published to Github Packages. Downloading the binary of the gem into vendor cache resolves the issue, but that is definitely a suboptimal solution. We’ve transitioned from Gemfury to Github Packages, and I’ve seen had an issue like this occur on Gemfury. Very confusing.</p>

**Comment by michaelwebb76** [comment]
<p>We’ve given up on Github Packages now and gone back to Gemfury. I think this one needs a bit more baking before it’s ready. We’re super keen to come back and use it once it’s done.</p>

**Comment by jdwright** [comment, marked as answer]
<p>I believe I had the same issue as the original post, in that the command line given by github to install the gem gave that error.  As far as I can tell, gem install does not use the credentials file.  But the command worked when I changed the github URL to include user name and token, as per basic authentication.</p>
<p>–source “<a>https://{username}:{token}@rubygems.pkg.github.com/</a>…”</p>

**Comment by jer-k** [comment]
<p>Finally circled back to this and I’m accepting this as the solution. It is not optimal, but it works for now!</p>
<p>Also just don’t forget to remove the <code>{}</code> around the input of your username and token. Was blankly staring at a non descriptive error like ‘why isn’t this working’ for 5 minutes…</p>

**Comment by asbjornu** [comment]
<p>Thanks for this. I’ve filed a <a href="https://github.com/rubygems/rubygems/issues/3467" rel="nofollow noopener">feature request with RubyGems</a> to have all commands support stored credentials consistently.</p>
